pdns/iputils.hh

```diff
@@ -369,6 +370,313 @@ private:
   uint8_t d_bits;
 };
 
+/** Per-bit binary tree map implementation with <Netmask,T> pair.
+ *
+ * This is an binary tree implementation for storing attributes for IPv4 and IPv6 prefixes.
+ * The most simple use case is simple NetmaskTree<bool> used by NetmaskGroup, which only
+ * wants to know if given IP address is matched in the prefixes stored.
+ *
+ * This element is useful for anything that needs to *STORE* prefixes, and *MATCH* IP addresses
+ * to a *LIST* of *PREFIXES*. Not the other way round.
+ *
+ * You can store IPv4 and IPv6 addresses to same tree, separate payload storage is kept per AFI.
+ *
+ * To erase something copy values to new tree sans the value you want to erase.
+ *
+ * Use swap if you need to move the tree to another NetmaskTree instance, it is WAY faster
+ * than using copy ctor or assigment operator, since it moves the nodes and tree root to
+ * new home instead of actually recreating the tree.
+ *
+ * Please see NetmaskGroup for example of simple use case. Other usecases can be found
+ * from GeoIPBackend and Sortlist, and from dnsdist.
+ */
+template <typename T>
+class NetmaskTree {
+public:
+  typedef Netmask key_type;
+  typedef T value_type;
+  typedef std::pair<key_type,value_type> node_type;
+  typedef size_t size_type;
+
+private:
+  /** Single node in tree, internal use only.
+    */
+  class TreeNode : boost::noncopyable {
+  public:
+     TreeNode(int bits) noexcept : d_bits(bits) {
+     }
+
+     //<! Makes a left node with one more bit than parent
+     TreeNode* make_left() {
+       if (!left) {
+         left = unique_ptr<TreeNode>(new TreeNode(d_bits+1));
+         left->parent = this;
+       }
+       return left.get();
+     }
+
+     //<! Makes a right node with one more bit than parent
+     TreeNode* make_right() {
+       if (!right) {
+         right = unique_ptr<TreeNode>(new TreeNode(d_bits+1));
+         right->parent = this;
+       }
+       return right.get();
+     }
+
+     unique_ptr<TreeNode> left;
+     unique_ptr<TreeNode> right;
+     TreeNode* parent;
```

Comment:
it seems `parent` is currently unused; maybe remove it, otherwise initialize it: `TreeNode* parent = nullptr`. Maybe pass as optional constructor argument instead of `x->parent = this` in `make_*` ?